charmstore/fakeclient.go

```diff
@@ -0,0 +1,582 @@
+	// Get returns the charm referenced by charmURL.
+	Get(charmURL *charm.URL) (charm.Charm, error)
+
+	// GetBundle returns the bundle referenced by bundleURL.
+	GetBundle(bundleURL *charm.URL) (charm.Bundle, error)
+
+	// // GetResource returns the resource referenced by resourceURL.
+	// GetResource(resourceURL *charm.URL, name string, revision int) (csclient.ResourceData, error)
+
+	// ResolveWithChannel resolves an entity to its canonical form, providing the series that
+	// the entity supports as a slice of strings.
+	Resolve(entityURL *charm.URL) (canonicalURL *charm.URL, supportedSeries []string, err error)
+
+	// ResolveWithChannel resolves an entity to its canonical form, providing the series that
+	// the entity supports as a slice of strings. Unlike Resolve, ResolveWithChannel includes the
+	// entity's current channel in the result.
+	ResolveWithChannel(*charm.URL) (*charm.URL, params.Channel, []string, error)
+}
+
+// NoopProgress implements csclient.Progress
+type NoopProgress struct{}
+
+func (NoopProgress) Start(uploadId string, expires time.Time) {}
+func (NoopProgress) Transferred(total int64)                  {}
+func (NoopProgress) Error(err error)                          {}
+func (NoopProgress) Finalizing()                              {}
+
+var _ csclient.Progress = (*NoopProgress)(nil)
+
+func NewProgress() csclient.Progress {
+	return NoopProgress{}
+}
+
+// Datastore is a small in-memory key/value store that can be used
+// for faking HTTP calls.
+type Datastore struct {
+	data map[string]interface{}
+}
+
+// NewDatastore returns an empty Datastore
+func NewDatastore() *Datastore {
+	return &Datastore{make(map[string]interface{})}
+}
+
+// Get retrieves data at path. If nothing exists at path,
+// an error satisfying errors.IsNotFound is returned.
+func (d *Datastore) Get(path string, data interface{}) error {
+	current := d.data[path]
+	if current == nil {
+		return errors.NotFoundf(path)
+	}
+	data = current
+	return nil
+}
+
+// Put stores data at path. It will be accessible later via Get.
+// Data already at path will is overwritten and no
+// revision history is saved.
+func (d *Datastore) Put(path string, data interface{}) error {
+	d.data[path] = data
+	return nil
+}
+
+// PutReader data from a an io.Reader at path. It will be accessible later via Get.
+// Data already at path will is overwritten and no
+// revision history is saved.
+func (d *Datastore) PutReader(path string, data io.Reader) error {
+	buffer := []byte{}
+	_, err := data.Read(buffer)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return d.Put(path, buffer)
+}
+
+// FakeClient is a stand-in charmstore.Client
+type FakeClient struct {
+	repo *Repository
+}
+
+var _ csWrapper = (*FakeClient)(nil)
+
+func NewFakeClient(repo *Repository) *FakeClient {
+	if repo == nil {
+		repo = NewRepository()
+	}
+	return &FakeClient{repo}
+}
+
+// Get retrieves data from path
+func (c FakeClient) Get(path string, value interface{}) error {
```

Comment:
Are all these methods needed for Deploy? If not, can we save them for another branch?

Reply:
Not for deploy specifically - they're there to satisfy what's defined within `charmrepo/csclient.Client`. I can happily whittle these down.